codechecker_lib/session_manager.py

```diff
@@ -0,0 +1,299 @@
+
+    @staticmethod
+    def calc_persistency_hash(client_addr, auth_string):
+        '''Calculates a more secure persistency hash for the session. This
+        persistency hash is intended to be used for the "session recycle"
+        feature to prevent NAT endpoints from accidentally getting each
+        other's session.'''
+        return hashlib.sha256(auth_string + "@" + client_addr + ":" +
+                              _Session.__initial_salt).hexdigest()
+
+    def __init__(self, client_addr, token, phash):
+        self.client = client_addr
+        self.token = token
+        self.persistent_hash = phash
+        self.last_access = datetime.now()
+
+    def still_valid(self, do_revalidate=False):
+        '''Returns if the session is still valid, and optionally revalidates
+        it. A session is valid in its soft-lifetime.'''
+        if (datetime.now() - self.last_access).total_seconds() <= \
+                session_lifetimes["soft"] \
+                and (datetime.now() - self.last_access).total_seconds() <= \
+                session_lifetimes["hard"]:
+            # If the session is still valid within the "reuse enabled" (soft)
+            # past and the check comes from a real user access, we revalidate
+            # the session by extending its lifetime --- the user retains their
+            # data.
+            if do_revalidate:
+                self.revalidate()
+
+            # The session is still valid if it has been used in the past
+            # (length of "past" is up to server host)
+            return True
+
+        # If the session is older than the "soft" limit,
+        # the user needs to authenticate again.
+        return False
+
+    def still_reusable(self):
+        '''Returns whether the session is still reusable, ie. within its
+        hard lifetime: while a session is reusable, a valid authentication
+        from the session's user will return the user to the session.'''
+        return (datetime.now() - self.last_access).total_seconds() <= \
+            session_lifetimes["hard"]
+
+    def revalidate(self):
+        if self.still_reusable():
+            # A session is only revalidated if it has yet to exceed its
+            # "hard" lifetime. After a session hasn't been used for this
+            # timeframe, it can NOT be resurrected at all --- the user needs
+            # to log in into a brand-new session
+            self.last_access = datetime.now()
+
+
+class sessionManager:
+    __valid_sessions = []
+    __logins_since_prune = 0
+
+    def __init__(self):
+        LOG.debug('Loading session config')
+
+        session_cfg_file = os.path.join(os.environ['CC_PACKAGE_ROOT'],
+                                        "config", "session_config.json")
+        LOG.debug(session_cfg_file)
+        with open(session_cfg_file, 'r') as scfg:
+            scfg_dict = json.loads(scfg.read())
+
+        if not scfg_dict["authentication"]:
```

Comment:
Actually the next line ```scfg_dict["authentication"] = {'enabled': False}``` could be moved up before loading the config json to create a dict where by default the authentication is disabled. In that case we do not have to check it later. A simple dict update```scfg_dict.update(scfg_from_json)``` could be used to overwrite the default values with the values from the config file .